src/lightning/data/streaming/dataset.py

```diff
@@ -11,36 +11,135 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Any, Literal, Optional, Union
+from typing import Any, List, Literal, Optional, Union
 
-from torch.utils.data import Dataset
+import numpy as np
+from torch.utils.data import IterableDataset
 
+from lightning.data.datasets.env import _DistributedEnv, _WorkerEnv
 from lightning.data.streaming import Cache
+from lightning.data.streaming.item_loader import BaseItemLoader
+from lightning.data.streaming.sampler import ChunkedIndex
 
 
-class StreamingDataset(Dataset):
+class StreamingDataset(IterableDataset):
```

Comment:
Just a check if I understand well: the change in the base class is for the dataloader to consume this as an iterable dataset (using `__iter__`), but `__len__` and `__getitem__` are also implemented. These are convenience methods for users at this point, since the dataloader will not use them. Which also means a custom Sampler won't be used.

But at the same time, `__next__` calls them, so it's a mapped dataset with an iterable dataset facade. What was the rationale for this change?

Reply:
Users don't know about IterableDataset and they don't need to. How we handle the sampling under the hood is up to us.
However, from the user perceptive, this behaves exactly as a map based dataset with a simple API. As a user, I just need to override the `__getitem__` method and voila.

```python
class ImageNetStreamingDataset(StreamingDataset):

    def __init__(self, name, version):
        super().__init__(name=name, version=version)
        self.transform = T.Compose([
            T.RandomResizedCrop((224, 224), antialias=True),
        ])

    def __getitem__(self, index):
        # Note: If torchvision is installed, we return a tensor image instead of a pil image as it is much faster. 
        img, class_index = super().__getitem__(index) # <- Whatever you returned from the DatasetOptimizer prepare_item method.
        return self.transform(to_rgb(img) / 255.), int(class_index)  
```

Reply:
This change has some implications for distributed sampling, because now the responsibility of sampling is with the dataset implementation, it also needs to include the dataloader-workers in the calculation, otherwise we end up with data duplication. Can we add a test that the dataset behaves correctly and equivalently to a regular dataset/dataloader with distributed sampler + multiple dataloader workers?

Reply:
Yes. But this is something we handle. So we just need to make this solid.